Decode JWT token for validation

roshii opened the pull request:
Implement `jmclient.auth` module to manage JWT.

Upon successful authentication (e.g. unlock wallet), response includes both a `token` and a `refresh_token`. The former can be used for call authentication, valid for 30 min. After expiration, user can call `/token/refresh` endpoint with his expired access token in header and refresh token in POST call payload to get both a new access and refresh token. Refresh token is valid for 4 hours.

Anytime a new access token is issued, refresh token signature key is re-initialized, invalidating any previously issued token.
Tokens are scoped to a specific `wallet_name` and a generic `walletrpc` category, and should allow future upgrades such as authorization granularity. 

Fixes https://github.com/JoinMarket-Org/joinmarket-clientserver/issues/1297

roshii commented:
Tests are passing locally but require Wegzeug<2.3.0 

Pending https://github.com/JoinMarket-Org/joinmarket-clientserver/pull/1479 resolution to have CI test passing and dependencies updated

roshii commented:
Now thinking that I did not consider the case when the token expires: how will client request a new token and authenticate itself?
To be continued

AdamISZ commented:
> Add `cookie_secret_key` and `cookie_signature_algorithm` to `JMWalletDaemon`. Decode token with set secret key, automatically validating expiry time with 10 s tolerance. Simplify `set_token` method by removing wallet_name parameter, reading it from property instead
> 
> Fixes #1297

Thanks a lot for working on this, it was badly needed.

roshii commented:
> Now thinking that I did not consider the case when the token expires: how will client request a new token and authenticate itself?

After analysis, I think token issuance responsibility should be moved to RPC clients. I did start working on an implementation in which a JWT (public) key parameter can be added to `JMWalletDaemon`. This key can later be used to verify token signature, would a token be expired, client will received a `409` which they have to deal with by issuing a new token.
For the sake of compatibility, if a secret is not provided `JMWalletDaemon` would create a token with a random secret, with 30 min validity a no way to refresh

AdamISZ commented:
> > I think token issuance responsibility should be moved to RPC clients.
> 
> Maybe not a good idea... I will rather refine server side implementation and address access token refresh issue with refresh tokens as described [here](https://auth0.com/blog/refresh-tokens-what-are-they-and-when-to-use-them/).

From a brief read, it does look like an interesting idea, though rather complex. Are you thinking about something like this? https://stackoverflow.com/a/71900761 .. if so, how would it work client-side? I didn't have a clear idea about this expiration issue myself, when we first (half-) implemented this.

roshii commented:
> I think token issuance responsibility should be moved to RPC clients.

Maybe not a good idea... I will rather refine server side implementation and address access token refresh issue with refresh tokens as described [here](https://auth0.com/blog/refresh-tokens-what-are-they-and-when-to-use-them/).

roshii commented:
Idea is as follows, for a typical flow:

1. Client unlock wallet -> receives an access token and a refresh token in the payload, and stores refresh token for later use
2. Client access authenticated endpoint with expired access token -> receives an error response with "invalid_token" error code
3. Client makes a refresh request with its expired access token and its refresh token, the latter in the payload -> server validates access token without expiration time verification, validates refresh token, and finally sends a new access and refresh token to client

AdamISZ commented:
Yes that's approximately what I understood, it is a bit complex, and I guess we don't get out of the box with the PyJWT package right? But from a logic perspective it seems entirely fine.

roshii commented:
Refresh logic requires a custom implementation indeed, PyJWT will solely help validating tokens

roshii commented:
I have just refined token validation implementation, it should now be straight forward to add a refresh token endpoint.

roshii commented:
Updated PR comment

AdamISZ commented:
Looks like really good work, I haven't yet found any issues in testing it out. I'll keep looking at it for a while.

You need to update https://github.com/JoinMarket-Org/joinmarket-clientserver/blob/master/docs/api/wallet-rpc.yaml with the new `/refresh` endpoint (and also update other endpoints like /unlock with the new `refresh_token` field). I'd also add some details in the `/refresh` endpoint that gives clear instructions to client coders on how to handle the refresh operation.

It really does seem a shame that we need such a large amount of code here to do stuff that really feels like it should be handled by an underlying library ... but it's certainly better to have sensible logic instead of basically broken (or, charitably "very incomplete") logic for tokens that we had before.

AdamISZ commented on jmclient/jmclient/auth.py:
```diff
@@ -0,0 +1,94 @@
+
+class JMSessionValidity:
+    access = datetime.timedelta(minutes=30)
+    refresh = datetime.timedelta(hours=4)
```
Nit: Questionable whether there's any point in a class here, if we just have two hardcoded values.

roshii replied:
Questionable indeed, I did wonder myself and ended up using a class for he following reasons.

1. Both values relates to a session validity, where each cannot exits without the other. It is the same object imho.
2. It allows to programmatically access one or the other value with `getattr(self.validity, token_type)`
3. It is consistent with `JMTokenAuthority.signature` attribute

AdamISZ commented on jmclient/jmclient/auth.py:
```diff
@@ -0,0 +1,94 @@
+
+    def issue(self, scope: str = "walletrpc") -> dict:
+        """Issue a new access and refresh token for said scope.
+        Previously issued refresh token and invalidated.
```
Nit: "and invalidated". "is invalidated"?

AdamISZ commented on jmclient/jmclient/auth.py:
```diff
@@ -0,0 +1,94 @@
+            algorithm=self.signature.algorithm,
+        )
+
+    def issue(self, scope: str = "walletrpc") -> dict:
```
Nit: perhaps default string should include 'joinmarket' in it, to specify scope better.

roshii replied:
I wondered as well. I ended up with the "shortened version" considering scope is encrypted and solely readable by `JMTokenAuthority`, it is internal only. We could skip it entirely at this stage but I thought it was good to leave a door open, would we need different authorization level in the future.

AdamISZ commented on jmclient/jmclient/websocketserver.py:
```diff
@@ -30,29 +31,30 @@ def connectionLost(self, reason):
         """
         WebSocketServerProtocol.connectionLost(self, reason)
         self.factory.unregister(self)
+        self.active_session = True
```
Why is this here? I don't think we want the session to to be set 'active' when the connection is lost?

roshii replied:
I can't recall why I did that but it makes no sense indeed, reverting.

AdamISZ commented:
OK should be ready after adding yaml and squashing.

roshii commented:
Updated OpenAPI doc and squashed

roshii commented:
> It really does seem a shame that we need such a large amount of code here to do stuff that really feels like it should be handled by an underlying library ... but it's certainly better to have sensible logic instead of basically broken (or, charitably "very incomplete") logic for tokens that we had before.

The issue is with the authorization logic that cannot really be delegated to a library imho. Considering every application will have different authorization needs, importing a "flexible library" would add many more lines of code than implementing a custom logic, which we do here with less than hundred lines.

AdamISZ commented on docs/api/wallet-rpc.yaml:
```diff
@@ -10,6 +10,26 @@ servers:
+        - bearerAuth: []
+      summary: refresh an access token
+      operationId: refreshtoken
+      description: Give a refresh token and get back both an access and refresh token. Access token, valid for 30 min, must be used for authenticated endpoints. refreshtoken endpoint can be called with an expired access token and a refresh token, valid 4 hours after issuance. The newly issued tokens must be used in subsequent calls since operation invalidates previously issued tokens.
```
> "valid 4 hours after issuance" means, if taken literally, that it is valid *4 or more* hours after issuance (and though obviously that is not what you mean, such a function does exist in JWT). Simplest change is "valid **for** 4 hours after ..".

AdamISZ commented:
@theborakompanioni looking at the doc in the yaml, do you find it easy to understand how to change the client code to do this?

@roshii 

... perhaps write a single paragraph to describe what the client code should do, like "on initially creating or unlocking an existing wallet, store both the refresh and access tokens, the latter is valid for only 30 minutes while the former is for 4 hours, so on some timescale (ideally less than 30 minutes for most simplicity), use /refresh to get a new token regularly" .. that's probably not exact right but something along those lines.

@roshii 

I did a bit more checking for obvious footguns like algorithm "none" but all looking good.

One gripe I have about the code is,why use `getattr`? It only makes it a little harder than it needs to be, to find usages of variables. We have another place with that (in WalletService) and I recall at least one person complained about it. Minor gripe, for sure, it doesn't really matter.

roshii commented:
> perhaps write a single paragraph to describe what the client code should do

I did try to keep it concise, maybe too much. I shall update description. Thanks

> One gripe I have about the code is,why use `getattr`?

I can easily modify code with something as follows, which even code even shorter. 

```
SESSION_VALIDITY = {
    access: datetime.timedelta(minutes=30),
    refresh: datetime.timedelta(hours=4),
}
SIGNATURE_ALGORITHM = "HS256"


def get_random_key(size: int = 16) -> str:
    """Create a random key has an hexadecimal string."""
    return bintohex(os.urandom(size))


class JMTokenAuthority:
    """Manage authorization tokens."""

    def __init__(self, wallet_name: Optional[str] = None):
        self.signature_key = {
            "access": get_random_key(),
            "refresh": get_random_key(),
        }
        self.wallet_name = wallet_name

    def verify(
        self,
        token: str,
        scope: str = "walletrpc",
        *,
        is_refresh: bool = False,
        verify_exp: bool = True,
    ):
        """Verify JWT token.

        Token must have a valid signature and its scope must contain both scopes in
        arguments and wallet_name property.
        """
        token_type = "refresh" if is_refresh else "access"
        claims = jwt.decode(
            token,
            self.signature_key[token_type],
            algorithms=SIGNATURE_ALGORITHM,
            options={"verify_exp": verify_exp},
            leeway=10,
        )
        token_claims = set(claims.get("scope", []).split())
        if not set(scope.split()) | {self.wallet_name} <= token_claims:
            raise InvalidScopeError

    def _issue(self, scope: str, token_type: str) -> str:
        return jwt.encode(
            {
                "exp": datetime.datetime.utcnow() + SESSION_VALIDITY[token_type],
                "scope": f"{scope} {self.wallet_name}",
            },
            self.signature_key[token_type],
            algorithm=SIGNATURE_ALGORITHM,
        )

    def issue(self, scope: str = "walletrpc") -> dict:
        """Issue a new access and refresh token for said scope.
        Previously issued refresh token is invalidated.
        """
        self.signature_key["refresh"] = get_random_key()
        return {
            "token": self._issue(scope, "access"),
            "refresh_token": self._issue(scope, "refresh"),
        }

    def reset(self):
        """Invalidate all previously issued tokens by creating new signature keys."""
        self.signature_key = {
            "access": get_random_key(),
            "refresh": get_random_key(),
        }
```

roshii commented:
> perhaps write a single paragraph to describe what the client code should do

@AdamISZ @theborakompanioni would that be better?

Give a refresh token and get back both an access and refresh token. On initially creating, unlocking or recovering an wallet, store both the refresh and access tokens, the latter is valid for only 30 minutes (must be used for any authenticated call) while the former is for 4 hours (can only be used in the refresh request parameters). Use /token/refresh endpoint on a regular basis to get new access and refresh tokens, ideally before access token expiration to avoid unauthorized response from authenticated endpoint and in any case, before refresh token expiration. The newly issued tokens must be used in subsequent calls since operation invalidates previously issued tokens.

AdamISZ commented:
> @AdamISZ @theborakompanioni would that be better?

Give a refresh token and get back both an access and refresh token. On initially creating, unlocking or recovering an wallet, store both the refresh and access tokens, the latter is valid for only 30 minutes (must be used for any authenticated call) while the former is for 4 hours (can only be used in the refresh request parameters). Use /token/refresh endpoint on a regular basis to get new access and refresh tokens, ideally before access token expiration to avoid unauthorized response from authenticated endpoint and in any case, before refresh token expiration. The newly issued tokens must be used in subsequent calls since operation invalidates previously issued tokens.

That looks perfect to me ("an wallet" -> a wallet or any wallet)

AdamISZ commented:
> > perhaps write a single paragraph to describe what the client code should do
> 
> I did try to keep it concise, maybe too much. I shall update description. Thanks
> 
> > One gripe I have about the code is,why use `getattr`?
> 
> I can easily modify code with something as follows, which even code even shorter.
> 
> ```
> SESSION_VALIDITY = {
>     access: datetime.timedelta(minutes=30),
>     refresh: datetime.timedelta(hours=4),
> }
> SIGNATURE_ALGORITHM = "HS256"
> 
> 
> def get_random_key(size: int = 16) -> str:
>     """Create a random key has an hexadecimal string."""
>     return bintohex(os.urandom(size))
> 
> 
> class JMTokenAuthority:
>     """Manage authorization tokens."""
> 
>     def __init__(self, wallet_name: Optional[str] = None):
>         self.signature_key = {
>             "access": get_random_key(),
>             "refresh": get_random_key(),
>         }
>         self.wallet_name = wallet_name
> 
>     def verify(
>         self,
>         token: str,
>         scope: str = "walletrpc",
>         *,
>         is_refresh: bool = False,
>         verify_exp: bool = True,
>     ):
>         """Verify JWT token.
> 
>         Token must have a valid signature and its scope must contain both scopes in
>         arguments and wallet_name property.
>         """
>         token_type = "refresh" if is_refresh else "access"
>         claims = jwt.decode(
>             token,
>             self.signature_key[token_type],
>             algorithms=SIGNATURE_ALGORITHM,
>             options={"verify_exp": verify_exp},
>             leeway=10,
>         )
>         token_claims = set(claims.get("scope", []).split())
>         if not set(scope.split()) | {self.wallet_name} <= token_claims:
>             raise InvalidScopeError
> 
>     def _issue(self, scope: str, token_type: str) -> str:
>         return jwt.encode(
>             {
>                 "exp": datetime.datetime.utcnow() + SESSION_VALIDITY[token_type],
>                 "scope": f"{scope} {self.wallet_name}",
>             },
>             self.signature_key[token_type],
>             algorithm=SIGNATURE_ALGORITHM,
>         )
> 
>     def issue(self, scope: str = "walletrpc") -> dict:
>         """Issue a new access and refresh token for said scope.
>         Previously issued refresh token is invalidated.
>         """
>         self.signature_key["refresh"] = get_random_key()
>         return {
>             "token": self._issue(scope, "access"),
>             "refresh_token": self._issue(scope, "refresh"),
>         }
> 
>     def reset(self):
>         """Invalidate all previously issued tokens by creating new signature keys."""
>         self.signature_key = {
>             "access": get_random_key(),
>             "refresh": get_random_key(),
>         }
> ```

Yes, that seems great

roshii commented:
- Implemented suggested changes
- Rebased
- Corrected YAML syntax error on `yieldgenreport` description

roshii commented:
Fix tests (without `getattr`)

theborakompanioni commented:
> @theborakompanioni looking at the doc in the yaml, do you find it easy to understand how to change the client code to do this?

Yep, I think it is quite straight-forward. Should be doable. Nice work.
However, doing it in a backward compatible way will be more tricky - but also manageable (i.e. if `refresh_token` is not present, keep old logic).
Could not find the mentioned durations anywhere (30min for access, 4h for refresh token) - I assume these are the default values of the library? Jam does not decode the token currently (as has not been necessary yet) - it must be done now to get the actual expiry values, correct?

AdamISZ commented:
> However, doing it in a backward compatible way will be more tricky - but also manageable (i.e. if refresh_token is not present, keep old logic).

I didn't think about this, but you're right in that, a user might have an old `joinmarketd` in their backend with a new version of the client (e.g. JAM). Indeed we should endeavour to make that still work fine, since it's possible so why not. Maybe @roshii *did* think about this and I just didn't notice. Will take a look.

> Could not find the mentioned durations anywhere (30min for access, 4h for refresh token) - I assume these are the default values of the library?

They are defined in `jmclient/auth.py`: https://github.com/JoinMarket-Org/joinmarket-clientserver/pull/1480/files#diff-a4264eb8f76fa45758d337287c2dda4d55283584e52adbb4d3c63c44ef316d4dR23 . Do you think we need these to be configurable?

>  Jam does not decode the token currently (as has not been necessary yet) - it must be done now to get the actual expiry values, correct?

Hmm, well for the expiry times it relates to the above (for now, no). For decoding in general, I think technically you do not need to decode, just store and send back with the refresh request, right?

roshii commented:
> Hmm, well for the expiry times it relates to the above (for now, no). For decoding in general, I think technically you do not need to decode, just store and send back with the refresh request, right?

Tokens can indeed be decoded by clients without signature verification. Expiry value is set in the `exp` claim. It is not required though, they can just be stored for later use.

@AdamISZ I realize here I made a mistake in one of my earlier comment: clients will be able to read scope (tokens are encoded not encrypted). Nevertheless, I do not see it necessary to add a `joinmarket` prefix.

> Indeed we should endeavour to make that still work fine, since it's possible so why not. Maybe @roshii did think about this and I just didn't notice. Will take a look.

We have two scenario here.

A. Old backend, new client (no refresh endpoint)

The logic is rather simple, implemented client side: if refresh_token present, then...

B. New backend, old client (short token validity)

I see two options here to facilitate transition:

1. Set defaults access token validity to say one day for access, two for refresh. This would apply to subsequent minor version update of joinmarket. In the next major release, set it back to low values.
2. Allow configurable token validity via environment variable (or config file?): clients not having implemented the refresh functionality could raise the value as needed.

theborakompanioni commented:
> They are defined in `jmclient/auth.py`: https://github.com/JoinMarket-Org/joinmarket-clientserver/pull/1480/files#diff-a4264eb8f76fa45758d337287c2dda4d55283584e52adbb4d3c63c44ef316d4dR23 . Do you think we need these to be configurable?

God, no! : D
Do I remember correctly that you once said it has "too many knobs already"?
Let's wait and see if someone really wants to customize these values.
However, invalidating the tokens when locking the wallet is a _must_ (is this done?).

> 
> > Jam does not decode the token currently (as has not been necessary yet) - it must be done now to get the actual expiry values, correct?
> 
> Hmm, well for the expiry times it relates to the above (for now, no). For decoding in general, I think technically you do not need to decode, just store and send back with the refresh request, right?

Jam can hardcode the defaults at first, but I'd rather be relying on the `exp` value later.


> B. New backend, old client (short token validity)
> 
> I see two options here to facilitate transition:
> 
>     1. Set defaults access token validity to say one day for access, two for refresh. This would apply to subsequent minor version update of joinmarket. In the next major release, set it back to low values.
> 
>     2. Allow configurable token validity via environment variable (or config file?): clients not having implemented the refresh functionality could raise the value as needed.

Being backward compatible is already hard enough as is. Forward compatibility as a requirement is unfeasible imho. It'd be nice to know the actual backend version (without "probing" for missing/existing endpoints) and display a warning in downstream clients instead. What do you think?

roshii commented:
> Let's wait and see if someone really wants to customize these values.

Fair enough. Hard coded values are best imho, this is not something users should be able to tweak. 

> However, invalidating the tokens when locking the wallet is a must (is this done?).

This is done yes.

> It'd be nice to know the actual backend version (without "probing" for missing/existing endpoints) and display a warning in downstream clients instead. What do you think?

I guess it doesn't arm to add a `/system-info` endpoint to return `joinmarket` version. Should be straight forward.

AdamISZ commented:
> I guess it doesn't arm to add a /system-info endpoint to return joinmarket version. Should be straight forward.

Agreed, that was an oversight. There is an API version number but I was never at all clear whether or what we should do with that. A simple /getinfo or similar including Joinmarket version (I think that is better than an extra joinmarketd version) makes sense.

AdamISZ commented:
@theborakompanioni it's a bit cheeky to ask  you as if it's your responsibility, but ... are you OK with this going in the next release? Afaict the main issue here is, since people's JM installations are separate from their JAM installations, it can cause potential problems for users if we don't ensure that it is backwards compatible.

I'll try to spend some time now looking into testing https://github.com/JoinMarket-Org/joinmarket-clientserver/commit/c5fe644f3dcdd2731c75ce92bd66d3dfee101771 and keep in mind the above conversation about compatibility.

theborakompanioni commented:
> @theborakompanioni it's a bit cheeky to ask you as if it's your responsibility, but ... are you OK with this going in the next release?

Thanks for asking - I very much welcome such improvements!


> Agreed, that was an oversight. There is an API version number but I was never at all clear whether or what we should do with that. A simple /getinfo or similar including Joinmarket version (I think that is better than an extra joinmarketd version) makes sense.
> 
> Maybe just call it /version.

I'd appreciate a `/getinfo` endpoint. In addition to the jm version, additional information can be included in the future. Jam would benefit from knowing the current blockheight joinmarket is aware of, as well as the network (and various other things I cannot think about right now). But that is for another issue.

Thank you guys!

AdamISZ commented:
> I'd appreciate a `/getinfo` endpoint. In addition to the jm version, additional information can be included in the future. Jam would benefit from knowing the current blockheight joinmarket is aware of, as well as the network (and various other things I cannot think about right now). But that is for another issue.

Yes seems very reasonable. More convenient to have one aggregated RPC call than have to do a ton of them to check different data. And add fields to the returned dict in later updates. As for the name, no point bikeshedding I guess, getinfo or sysinfo or whatever.

kristapsk commented:
> As for the name, no point bikeshedding I guess, getinfo or sysinfo or whatever.

`getinfo` seems the best name, it's how others call it - Core has `get***info` calls, CLN has `getinfo`.

AdamISZ commented:
See #1532 

Hopefully this addresses backwards compatibility concerns, as well as being something we can add to later.

AdamISZ commented on jmclient/test/test_wallet_rpc.py:
```diff
@@ -712,6 +718,88 @@ def process_get_seed_response(self, response, code):
+            ("expired", "expired", False),
+            ("invalid", "valid", False),
+        ]:
+            access_token, refresh_token, is_successful_response = "valid", "valid", True
```
@roshii I don't understand: this line is inside the `for` loop, but it resets the values of the loop variables so they're always the same?

roshii replied:
> this line is inside the `for` loop, but it resets the values of the loop variables so they're always the same?

I can remember creating the test without the `for` loop with this line, should be gone indeed...

AdamISZ commented on jmclient/test/test_wallet_rpc.py:
```diff
@@ -712,6 +718,88 @@ def process_get_seed_response(self, response, code):
+
+    def get_token(self, *, expired=False, invalid=False, refresh=False):
+        if invalid:
+            return jwt.encode({"scope": "invalid"}, "12345678")
```
@roshii Arguably a nit, but: to test that we don't accept invalid tokens is .. hard; it's not sufficient to give one obviously invalid string.

First, it'd be better to have a whole dataset of invalid token encodings, but that's asking a bit much. If we're only going to test 1 or a few invalid strings, I tend in the opposite direction: create a valid string and malleate one byte (or char).

AdamISZ replied:
Oh, and I forgot to also mention, you should definitely use a valid *scope* here, right?

roshii replied:
> Oh, and I forgot to also mention, you should definitely use a valid _scope_ here, right?

Right, valid _scope_ with invalid signature (and opposite) should be tested as well. I will use the proposed approach to create a few more test cases.

AdamISZ commented:
Two things that are not hugely needed but worth considering:

1. Add a note to https://github.com/JoinMarket-Org/joinmarket-clientserver/blob/master/docs/JSON-RPC-API-using-jmwalletd.md#rules-about-making-requests about the token refresh scheme.
2. Perhaps mention there the algorithm chosen for the token signing (not really needed for clients, but, just to doc it). Relatedly, consider adding a test case where the signing algorithm is the wrong one, to verify that it is rejected as expected.

AdamISZ commented on jmclient/jmclient/auth.py:
```diff
@@ -0,0 +1,85 @@
+        arguments and wallet_name property.
+        """
+        token_type = "refresh" if is_refresh else "access"
+        claims = jwt.decode(
```
Afaict from looking at the pyjwt code, this will throw `ExpiredSignatureError` if we passed the expiration. I think there's a value in catching that to bubble up an error message for the caller? So they can distinguish between a rejected token that is just invalid vs one that is expired?

Sometimes there's an argument in computer security for never telling the caller the reason for the failure, since it can allow probing attacks. That doesn't seem to apply here.

roshii commented:
@AdamISZ you'll have my comments shortly. We may want to modify a few things.

roshii commented:
> @roshii if you're not inclined to make further updates, I'm pretty much fine with merging as-is. (Minor exception [here](https://github.com/JoinMarket-Org/joinmarket-clientserver/pull/1480#discussion_r1294000181) but whatever, I can edit that later if needed).

Reading [OAuth 2.1 draft specs](https://www.ietf.org/archive/id/draft-ietf-oauth-v2-1-09.html#name-token-endpoint) but I came to realize that my implementation differs from it on a few point.

1. Endpoint should be `/token` - not `/token/refresh`
2. Endpoint should accept `POST` method only
3. `refresh` _grant_ should be specified in HTTP request content
4. Response should have an `access_token` parameter but since we already use `token` I propose to keep the latter.

Since changes are rather straight forward, I would like to align to standards before merging.

AdamISZ commented:
> refresh grant should be specified in HTTP request content

Could you explain point 3? I don't understand what that means.

roshii commented:
> > refresh grant should be specified in HTTP request content
> 
> Could you explain point 3? I don't understand what that means.

An example will speak by itself:

- My implementation implies the following call:
`curl -X GET https://localhost/api/v1/token/refresh --data 'refresh_token=a7f1...'`

- OAuth recommends the following:
`curl -X POST https://localhost/api/v1/token --data 'grant_type=refresh&refresh_token=a7f1...'`

roshii commented:
- Rebased on `master`
- Addressed review comments
- Align to OAuth specs, with two exceptions (in response key to maintain compatibility with `joinmarket` implementation

AdamISZ commented:
I'll make it a priority to review it as I can over the next day or so. @theborakompanioni any feedback from you (or other JAM people) would be great.

AdamISZ commented on jmclient/jmclient/wallet_rpc.py:
```diff
@@ -582,6 +599,57 @@ def preflight(self, request):
+
+            def _mkerr(err, description=""):
+                return make_jmwalletd_response(
+                    request, status=400, message=err, error_description=description
```
Shouldn't some (or all?) of the places where this is called, be using 401 and not 400?

roshii replied:
Authorization happens when verifying bearer token, if this verification fails a 401 will be returned.
If refresh token verification fails, it is a request error, 400, in line with https://datatracker.ietf.org/doc/html/draft-ietf-oauth-v2-1-08#name-error-response

AdamISZ replied:
Read and understood, thanks.

AdamISZ commented:
Have reviewed the update and I believe it all makes sense as far as I have checked. Also tested it out, though only using the test suite. I just have that one question above for now, though I will probably try to check a couple more things.

AdamISZ commented:
@roshii 

OK so I tested this in as realistic a way as I could (using POSTMAN) to call `/token` and I found an error:

In the `refresh()` method in JMWalletDaemon, you are calling `get_POST_body()` twice, but the first one has already read the byestream object, so the second call returns `b''` from the `request.content.read()`. Hence calling `/token` results in an error (because `get_POST_body` just returns `False` and you subscript that return value).

I understand that the intention was to have it be dynamic, i.e. that the `grant_type` field defines the key name of the field containing the actual "thing" which is currently just `"refresh_token"` and nothing else. That doesn't quite fit with how `get_POST_body was written. So for now I'm going to take the "executive decision" to merge this then add the small patch that fixes that error by just only allowing a "refresh_token" field; but you're more than welcome to just propose another change if you like, after.

AdamISZ commented:
@theborakompanioni please see the follow up commit directly after ( a847df9e1dc4c4fcff6ffc30610499e3534fd07e ), and let us know how this goes for JAM clients, as I can well imagine this is a non trivial extra bit of coding client side.

roshii commented:
> I understand that the intention was to have it be dynamic, i.e. that the `grant_type` field defines the key name of the field containing the actual "thing" which is currently just `"refresh_token"` and nothing else. That doesn't quite fit with how `get_POST_body was written. So for now I'm going to take the "executive decision" to merge this then add the small patch that fixes that error by just only allowing a "refresh_token" field; but you're more than welcome to just propose another change if you like, after.

Thanks for the fix. I will address the dynamic behavior of this method while implementing authentication/credentials (https://github.com/JoinMarket-Org/joinmarket-clientserver/issues/3) which will most likely need to manage another grant type.

AdamISZ commented:
> > I understand that the intention was to have it be dynamic, i.e. that the `grant_type` field defines the key name of the field containing the actual "thing" which is currently just `"refresh_token"` and nothing else. That doesn't quite fit with how `get_POST_body was written. So for now I'm going to take the "executive decision" to merge this then add the small patch that fixes that error by just only allowing a "refresh_token" field; but you're more than welcome to just propose another change if you like, after.
> 
> Thanks for the fix. I will address the dynamic behavior of this method while implementing authentication/credentials (#3) which will most likely need to manage another grant type.

Right, sounds good. I realized after pushing it that I was being a bit braindead, as well as any sophisticated way of doing it, there is a very simply way of doing it: deep copy the `request` object :) But right, fold it in with another change.

theborakompanioni commented:
> @theborakompanioni please see the follow up commit directly after ( [a847df9](https://github.com/JoinMarket-Org/joinmarket-clientserver/commit/a847df9e1dc4c4fcff6ffc30610499e3534fd07e) ), and let us know how this goes for JAM clients, as I can well imagine this is a non trivial extra bit of coding client side.

Will immediately start after the new Jam version for v0.9.10 is released. :pray:

theborakompanioni commented:
I am currently implementing this for Jam.
First thing noticeable is that there seems to be a problem with wallets having space (` `) in it, e.g. `import test.jmdat`.

This is the result of a recovery test (`/recover` endpoint):

`import.jmdat`:
![image](https://github.com/JoinMarket-Org/joinmarket-clientserver/assets/3358649/b4a0ecfc-dced-44c4-a546-06fe665efb44)

```json
scope: "test.jmdat walletrpc import.jmdat"
```

`import test.jmdat`:
![image](https://github.com/JoinMarket-Org/joinmarket-clientserver/assets/3358649/99563412-b1fb-495a-bf46-55c7824b8b53)

```json
scope: "test.jmdat walletrpc import test.jmdat import.jmdat"
```

Note: Existing wallets are `test.jmdat`, `import.jmdat` and `import test.jmdat`.

roshii commented:
> I am currently implementing this for Jam.
> First thing noticeable is that there seems to be a problem with wallets having space (` `) in it, e.g. `import test.jmdat`.
> 
> This is the result of a recovery test (`/recover` endpoint):
> 
> `import.jmdat`:
> ![image](https://github.com/JoinMarket-Org/joinmarket-clientserver/assets/3358649/b4a0ecfc-dced-44c4-a546-06fe665efb44)
> 
> ```json
> {
>  [...]
>   "scope": "test.jmdat walletrpc import.jmdat"
> }
> ```
> 
> `import test.jmdat`:
> ![image](https://github.com/JoinMarket-Org/joinmarket-clientserver/assets/3358649/99563412-b1fb-495a-bf46-55c7824b8b53)
> 
> ```json
> {
>  [...]
>   "scope": "test.jmdat walletrpc import test.jmdat import.jmdat"
> }
> ```
> 
> Note: Existing wallets are `test.jmdat`, `import.jmdat` and `import test.jmdat`.

Thanks for the info, I'll have a look at it shortly.

theborakompanioni commented:
fyi @roshii, I don't know if this is on purpose or a mistake. When the access token is not refreshed and a request is made with the expired token, the message in `www-authenticate` response header contains a stacktrace:
```
HTTP/1.1 401 Unauthorized
X-Powered-By: Express
transfer-encoding: chunked
connection: close
server: TwistedWeb/22.4.0
date: Wed Sep 20 2023 18:50:21 GMT-0700
www-authenticate: error_description="[Failure instance: Traceback: <class 'jmclient.wallet_rpc.InvalidToken'>: The access token provided is expired. /usr/local/lib/python3.9/dist-packages/twisted/web/server.py:292:render /usr/local/lib/python3.9/dist-packages/klein/_resource.py:204:render /usr/local/lib/python3.9/dist-packages/twisted/internet/defer.py:190:maybeDeferred /usr/local/lib/python3.9/dist-packages/klein/_resource.py:196:_execute --- <exception caught here> --- /usr/local/lib/python3.9/dist-packages/twisted/internet/defer.py:190:maybeDeferred /usr/local/lib/python3.9/dist-packages/klein/_app.py:129:execute_endpoint /usr/local/lib/python3.9/dist-packages/klein/_app.py:241:_f /usr/local/lib/python3.9/dist-packages/klein/_app.py:60:_call /src/jmclient/jmclient/wallet_rpc.py:707:session /src/jmclient/jmclient/wallet_rpc.py:412:check_cookie_if_present /src/jmclient/jmclient/wallet_rpc.py:400:check_cookie ]"
content-type: text/html
```

Except for the report above (space in wallet name), it seems to work fine. :muscle: 
Will keep reporting additional findings, should anything else come up.

roshii commented:
> fyi @roshii, I don't know if this is on purpose or a mistake. When the access token is not refreshed and a request is made with the expired token, the message in `www-authenticate` response header contains a stacktrace:
> ```
> HTTP/1.1 401 Unauthorized
> X-Powered-By: Express
> transfer-encoding: chunked
> connection: close
> server: TwistedWeb/22.4.0
> date: Wed Sep 20 2023 18:50:21 GMT-0700
> www-authenticate: error_description="[Failure instance: Traceback: <class 'jmclient.wallet_rpc.InvalidToken'>: The access token provided is expired. /usr/local/lib/python3.9/dist-packages/twisted/web/server.py:292:render /usr/local/lib/python3.9/dist-packages/klein/_resource.py:204:render /usr/local/lib/python3.9/dist-packages/twisted/internet/defer.py:190:maybeDeferred /usr/local/lib/python3.9/dist-packages/klein/_resource.py:196:_execute --- <exception caught here> --- /usr/local/lib/python3.9/dist-packages/twisted/internet/defer.py:190:maybeDeferred /usr/local/lib/python3.9/dist-packages/klein/_app.py:129:execute_endpoint /usr/local/lib/python3.9/dist-packages/klein/_app.py:241:_f /usr/local/lib/python3.9/dist-packages/klein/_app.py:60:_call /src/jmclient/jmclient/wallet_rpc.py:707:session /src/jmclient/jmclient/wallet_rpc.py:412:check_cookie_if_present /src/jmclient/jmclient/wallet_rpc.py:400:check_cookie ]"
> content-type: text/html
> ```
> 
> Except for the report above (space in wallet name), it seems to work fine. :muscle: 
> Will keep reporting additional findings, should anything else come up.

Thanks for the info. Headers should contain the following info, with no hint on class name.

`www-authenticate: error_description=The access token provided is expired.`

I'll take a look to have it done has it should.
